ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,284 @@
+	}
+	errorMultipleOfZonesCount = func(optimalTokenOwnership uint32, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: optimal token ownership must be a positive multiple of maximal allowed number of zones %d", token.prevToken, token.token, optimalTokenOwnership, maxZonesCount)
+	}
+	errorLowerAndUpperBoundModulo = func(optimalTokenOwnership uint32, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: lower and upper bounds must be congruent modulo maximal allowed number of zones %d", token.prevToken, token.token, optimalTokenOwnership, maxZonesCount)
+	}
+	errorDistanceBetweenTokensNotBigEnough = func(optimalTokenOwnership, ownership int, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: distance between lower and upper bound %d is not big enough", token.prevToken, token.token, optimalTokenOwnership, ownership)
+	}
+)
+
+type SpreadMinimizingConfig struct {
+	InstanceID string
+	Zone       string
+}
+
+type SpreadMinimizingTokenGenerator struct {
+	cfg        SpreadMinimizingConfig
+	instanceID int
+	zoneID     int
+	zones      []string
+	logger     log.Logger
+}
+
+func NewSpreadMinimizingTokenGenerator(cfg SpreadMinimizingConfig, zones []string, logger log.Logger) (*SpreadMinimizingTokenGenerator, error) {
+	sortedZones := make([]string, len(zones))
+	copy(sortedZones, zones)
+	if !slices.IsSorted(sortedZones) {
+		sort.Strings(sortedZones)
+	}
+	instanceID, err := parseInstanceID(cfg.InstanceID)
+	if err != nil {
+		return nil, err
+	}
+	zoneID, err := findZoneID(cfg.Zone, sortedZones)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenGenerator := &SpreadMinimizingTokenGenerator{
+		cfg:        cfg,
+		instanceID: instanceID,
+		zoneID:     zoneID,
+		zones:      sortedZones,
+		logger:     logger,
+	}
+	return tokenGenerator, nil
+}
+
+func parseInstanceID(instanceID string) (int, error) {
+	parts := instanceIDRegex.FindStringSubmatch(instanceID)
+	if len(parts) != 4 {
+		return 0, errorBadInstanceIDFormat(instanceID)
+	}
+	return strconv.Atoi(parts[3])
+}
+
+// findZoneID gets a zone name and a slice of sorted zones,
+// and return the index of the zone in the slice.
+func findZoneID(zone string, sortedZones []string) (int, error) {
+	if !slices.IsSorted(sortedZones) {
+		slices.Sort(sortedZones)
+	}
+	index := sort.SearchStrings(sortedZones, zone)
+	if index >= len(sortedZones) {
+		return 0, errorZoneNotValid(zone)
+	}
+	return index, nil
+}
+
+// generateFirstInstanceTokens calculates a set of tokens that should be assigned to the first instance (with id 0)
+// of the zone of the underlying instance.
+func (t *SpreadMinimizingTokenGenerator) generateFirstInstanceTokens() Tokens {
+	tokenDistance := (totalTokensCount / optimalTokensPerInstance / maxZonesCount) * maxZonesCount
```

Comment:
I think you will have to explain me why we need to do `/ maxZonesCount) * maxZonesCount` 🤔

Reply:
Similarly to the previous comment, in this approach tokens from one zone are always equal to each other modulo `maxZonesCount`. Therefore, I calculated tokenDistance as a multiple of `maxZonesCount`, so we are sure that the loop that calculates the actual token will satisfy that pattern.
With the current values of constants `totalTokensCount`, `optimalTokensPerInstance` and `maxZonesCount` (being $2^{32}$, $2^{9}$ and $2^{3}$) it is not necessary to do `/ maxZonesCount) * maxZonesCount`, but the check is there to guarante that if those values are changed, the property of multilpe of `maxZonesCount` is still guaranteed.

Reply:
[nit] Can you add a comment to explain it briefly, please?